### ISSUE TITLE
Minor improvements to benchmark combinations

### ISSUE THREAD

**Issue by snakefoot** [maciejw/logger-perf-tests#1]
This reduces the number of relevant benchmark combinations:
```c#
    [Flags]
    public enum FileMode
    {
        None = 0,
        KeepFileOpen = 1,
        Buffered = 2,
        Shared = 4,
        KeepFileOpenBuffered = KeepFileOpen | Buffered,
        KeepFileOpenShared = KeepFileOpen | Shared,
    }

    [RPlotExporter]
    [SimpleJob(RuntimeMoniker.NetCoreApp30)]
    public class SingleThreadBenchmarks
    {
        private SerilogTests serilogBaseTests;
        private NLogTests nLogTests;
        private Log4NetTests log4NetTests;

        [Params(FileMode.None, FileMode.KeepFileOpen, FileMode.KeepFileOpenBuffered, FileMode.KeepFileOpenShared)]
        public FileMode FileMode { get; set; }

        #region Setup
        [GlobalSetup(Target = nameof(Log4Net))]
        public void Log4NetGlobalSetup()
        {
            log4NetTests = new Log4NetTests(new Log4NetConfiguration { Buffered = FileMode.HasFlag(FileMode.Buffered), KeepFileOpen = FileMode.HasFlag(FileMode.KeepFileOpen), Shared = FileMode.HasFlag(FileMode.Shared) });
        }

        [GlobalSetup(Target = nameof(NLog))]
        public void NLogGlobalSetup()
        {
            nLogTests = new NLogTests(new NLogConfiguration { Buffered = FileMode.HasFlag(FileMode.Buffered), KeepFileOpen = FileMode.HasFlag(FileMode.KeepFileOpen), Shared = FileMode.HasFlag(FileMode.Shared) });
        }

        [GlobalSetup(Target = nameof(Serilog))]
        public void SerilogGlobalSetup()
        {
            serilogBaseTests = new SerilogTests(new SerilogConfiguration { Buffered = FileMode.HasFlag(FileMode.Buffered), Shared = FileMode.HasFlag(FileMode.Shared) });
        }

        ....
    }
}
```

This adds support for shared in log4net:

```c#
            log4net.Appender.FileAppender.LockingModelBase LockingModel = new log4net.Appender.FileAppender.MinimalLock();
            if (KeepFileOpen)
            {
                if (Shared)
                      LockingModel = new log4net.Appender.FileAppender.InterProcessLock();
                else
                      LockingModel = new log4net.Appender.FileAppender.ExclusiveLock();
            }
```

The NLog-config has some extra overhead:
- Unlike Serilog / log4net then NLog is told to render `${date}` dynamically for each logevent
- Unlike Serilog / log4net then NLog is told to lookup `${currentdir}` for each logevent

This NLog-FileTarget matches log4net/serilog:
```c#
            NLog.Targets.FileTarget fileTarget = new NLog.Targets.FileTarget("audit-log")
            {
                FileName = $"./nlog/audit-{DateTime.Now.ToString("yyyyMMddHHmm")}-latest.log",
                ArchiveFileName = $"./nlog/audit-{DateTime.Now.ToString("yyyyMMddHHmm")}-{{###}}.log",
```

The NLog-JsonLayout has the option to skip json-safe-encoding using `{ Encode = false }`:

```c#
            NLog.Layouts.JsonLayout jsonLayout = new NLog.Layouts.JsonLayout
            {
                IncludeAllProperties = true,
                Attributes = {
                    new NLog.Layouts.JsonAttribute("@t", "${longdate}") { Encode = false },
                    new NLog.Layouts.JsonAttribute("@l", "${level:upperCase=true}") { Encode = false },
                    new NLog.Layouts.JsonAttribute("@mt", "${message:raw=true}"),
                }
            };
```

**Comment by snakefoot** [comment]
My results on NetCoreApp3:

|  Method |             FileMode |        Mean |     Error |    StdDev | Ratio | RatioSD |
|-------- |--------------------- |------------:|----------:|----------:|------:|--------:|
|    NLog |                 None | 1,006.95 us | 20.072 us | 47.312 us | 40.15 |    2.34 |
| Log4Net |                 None | 1,023.78 us | 22.972 us | 67.732 us | 37.90 |    1.90 |
|         |                      |             |           |           |       |         |
| Serilog |         KeepFileOpen |    25.46 us |  0.405 us |  0.379 us |  1.00 |    0.00 |
|    NLog |         KeepFileOpen |    27.71 us |  0.544 us |  0.648 us |  1.09 |    0.03 |
| Log4Net |         KeepFileOpen |    71.16 us |  1.611 us |  1.978 us |  2.80 |    0.07 |
|         |                      |             |           |           |       |         |
| Serilog | KeepFileOpenBuffered |    16.84 us |  0.210 us |  0.186 us |  1.00 |    0.00 |
|    NLog | KeepFileOpenBuffered |    16.08 us |  0.321 us |  0.285 us |  0.95 |    0.02 |
| Log4Net | KeepFileOpenBuffered |    59.81 us |  1.045 us |  1.073 us |  3.56 |    0.08 |
|         |                      |             |           |           |       |         |
| Serilog | KeepFileOpenShared |    37.73 us |  0.723 us |  0.804 us |  1.00 |    0.00 |
|    NLog | KeepFileOpenShared |   561.12 us |  7.812 us |  6.926 us | 14.96 |    0.47 |
| Log4Net | KeepFileOpenShared |    80.72 us |  1.589 us |  3.024 us |  2.13 |    0.09 |

NLog has an issue with `KeepFileOpenShared` when combined with file-archiving on NetCore. It reopens the file for every logevent to check the filesize. This will hopefully be fixed in NLog 4.7

**Comment by maciejw** [comment]
thanks for letting me know, I'll fix few things you've mentioned, this looks much better. 
I want to make tests that are most similar to each other, so those point you've raised are very good.
I'll try to add some multi-threading tests also...

**Comment by snakefoot** [comment]
Happy to help. And thank you for highlighting the issue with "KeepFileOpenShared". Pull-requests have been merged for NLog 4.7 that will fix the performance issues raised.

**Comment by snakefoot** [comment]
My results using NLog 4.7 RC:

|  Method |       Runtime |             FileMode |         Mean |      Error |    StdDev |  Ratio | RatioSD |
|-------- |-------------- |--------------------- |-------------:|-----------:|----------:|-------:|--------:|
|    NLog |      .NET 4.8 |                 None |  9,220.03 us | 113.167 us | 94.499 us | 254.60 |   23.58 |
| Log4Net |      .NET 4.8 |                 None | 10,424.77 us | 118.437 us | 98.900 us | 287.63 |   23.84 |
|         |               |                      |              |            |           |        |         |
|    NLog | .NET Core 3.1 |                 None |    895.73 us |  16.048 us | 15.012 us |  34.56 |    0.80 |
| Log4Net | .NET Core 3.1 |                 None |    744.11 us |  14.417 us | 16.602 us |  28.60 |    0.80 |
|         |               |                      |              |            |           |        |         |
| Serilog |      .NET 4.8 |         KeepFileOpen |     35.12 us |   0.645 us |  0.572 us |   1.00 |    0.00 |
|    NLog |      .NET 4.8 |         KeepFileOpen |     36.11 us |   0.531 us |  0.497 us |   1.03 |    0.01 |
| Log4Net |      .NET 4.8 |         KeepFileOpen |    353.04 us |   1.181 us |  0.986 us |  10.04 |    0.16 |
|         |               |                      |              |            |           |        |         |
| Serilog | .NET Core 3.1 |         KeepFileOpen |     27.13 us |   0.255 us |  0.239 us |   1.00 |    0.00 |
|    **NLog** | **.NET Core 3.1** |         **KeepFileOpen** |     **26.54 us** |   0.347 us |  0.325 us |   0.98 |    0.02 |
| Log4Net | .NET Core 3.1 |         KeepFileOpen |     71.90 us |   1.030 us |  0.913 us |   2.65 |    0.04 |
|         |               |                      |              |            |           |        |         |
| Serilog |      .NET 4.8 | KeepFileOpenBuffered |     25.40 us |   0.465 us |  0.435 us |   1.00 |    0.00 |
|    NLog |      .NET 4.8 | KeepFileOpenBuffered |     22.59 us |   0.638 us |  0.682 us |   0.88 |    0.02 |
| Log4Net |      .NET 4.8 | KeepFileOpenBuffered |    336.10 us |   1.247 us |  1.042 us |  13.25 |    0.25 |
|         |               |                      |              |            |           |        |         |
| Serilog | .NET Core 3.1 | KeepFileOpenBuffered |     17.45 us |   0.270 us |  0.226 us |   1.00 |    0.00 |
|    **NLog** | **.NET Core 3.1** | **KeepFileOpenBuffered** |     **15.76 us** |   0.265 us |  0.235 us |   0.90 |    0.02 |
| Log4Net | .NET Core 3.1 | KeepFileOpenBuffered |     59.99 us |   0.392 us |  0.327 us |   3.44 |    0.05 |
|         |               |                      |              |            |           |        |         |
| Serilog |      .NET 4.8 |   KeepFileOpenShared |     39.79 us |   0.504 us |  0.471 us |   1.00 |    0.00 |
|    NLog |      .NET 4.8 |   KeepFileOpenShared |     43.03 us |   0.421 us |  0.394 us |   1.08 |    0.02 |
| Log4Net |      .NET 4.8 |   KeepFileOpenShared |    359.28 us |   3.187 us |  2.661 us |   9.02 |    0.11 |
|         |               |                      |              |            |           |        |         |
| **Serilog** | **.NET Core 3.1** |   **KeepFileOpenShared** |     **39.04 us** |   0.329 us |  0.307 us |   1.00 |    0.00 |
|    NLog | .NET Core 3.1 |   KeepFileOpenShared |     44.64 us |   0.863 us |  0.808 us |   1.14 |    0.03 |
| Log4Net | .NET Core 3.1 |   KeepFileOpenShared |     85.77 us |   1.962 us |  4.224 us |   2.12 |    0.10 |

Happy that KeepFileOpenShared for NLog 4.7 is no longer gives a huge performance hit, when combined with file-archive-logic.
